### DISCUSSION TITLE
Google OAuth access token issue - mismatching_state: CSRF Warning!

### DISCUSSION THREAD

**Discussion by alexkruWix** [apache/airflow#21066] [Q&A]
### Apache Airflow version

2.2.3 (latest released)

### What happened

When installing the new version of Airflow in our organization, and connecting it to Google OAuth, we encounter a problem after the initial setup - While the scheduler is working fine, we can not login into Airflow.

Once we select the Google account, we are redirected back to the login screen, and can not go into the UI.
In the debug logs, we see the following:
```
[2022-01-19 12:15:34,035] {views.py:615} DEBUG - Provider: google
[2022-01-19 12:15:34,035] {views.py:615} DEBUG - Provider: google
[2022-01-19 12:15:34,036] {views.py:628} DEBUG - Going to call authorize for: google
[2022-01-19 12:15:34,036] {views.py:628} DEBUG - Going to call authorize for: google
[2022-01-19 12:15:34,036] {base_app.py:155} DEBUG - Saving authorize data: <JSON_AUTHORIZE_DATA>
[2022-01-19 12:15:36,389] {views.py:658} DEBUG - Authorized init
[2022-01-19 12:15:36,389] {views.py:658} DEBUG - Authorized init
[2022-01-19 12:15:36,389] {views.py:666} ERROR - Error authorizing OAuth access token: mismatching_state: CSRF Warning! State not equal in request and response.
[2022-01-19 12:15:36,389] {views.py:666} ERROR - Error authorizing OAuth access token: mismatching_state: CSRF Warning! State not equal in request and response.
[2022-01-19 12:15:36,781] {views.py:615} DEBUG - Provider: None
[2022-01-19 12:15:36,781] {views.py:615} DEBUG - Provider: None
``` 

So far we tried:
* We set the secret_key config both on the airflow.cfg file, and as all the specified env vars, but we are still having the same issue.
* Verified that `cookie_samesite` is set to `Lax` and verified redirect_uri's in Google's client (based on this [thread](https://github.com/apache/airflow/issues/12812#issuecomment-738860503))
* Set `WTF_CSRF_ENABLED` to True
* Tried to change the secret_key to a "more simple" value

But nothing seems to work.
We are using the following `webserver_config.py`:
```
"""Default configuration for the Airflow webserver"""
import os

from airflow import configuration as conf
#from airflow.www.fab_security.manager import AUTH_DB

# from airflow.www.fab_security.manager import AUTH_LDAP
from airflow.www.fab_security.manager import AUTH_OAUTH
# from airflow.www.fab_security.manager import AUTH_OID
# from airflow.www.fab_security.manager import AUTH_REMOTE_USER


basedir = os.path.abspath(os.path.dirname(__file__))

#Note sure if this should stay - it doesnt exist in airflow 2?
SQLALCHEMY_DATABASE_URI = conf.get('core', 'SQL_ALCHEMY_CONN')

# Flask-WTF flag for CSRF
WTF_CSRF_ENABLED = True

# ----------------------------------------------------
# AUTHENTICATION CONFIG
# ----------------------------------------------------
# For details on how to set up each of the following authentication, see
# http://flask-appbuilder.readthedocs.io/en/latest/security.html# authentication-methods
# for details.

# The authentication type
# AUTH_OID : Is for OpenID
# AUTH_DB : Is for database
# AUTH_LDAP : Is for LDAP
# AUTH_REMOTE_USER : Is for using REMOTE_USER from web server
# AUTH_OAUTH : Is for OAuth
AUTH_TYPE = AUTH_OAUTH

# Uncomment to setup Full admin role name
AUTH_ROLE_ADMIN = 'Admin'

# Uncomment and set to desired role to enable access without authentication
# AUTH_ROLE_PUBLIC = 'Viewer'

# Will allow user self registration
AUTH_USER_REGISTRATION = True

# The recaptcha it's automatically enabled for user self registration is active and the keys are necessary
# RECAPTCHA_PRIVATE_KEY = PRIVATE_KEY
# RECAPTCHA_PUBLIC_KEY = PUBLIC_KEY

# Config for Flask-Mail necessary for user self registration
# MAIL_SERVER = 'smtp.gmail.com'
# MAIL_USE_TLS = True
# MAIL_USERNAME = 'yourappemail@gmail.com'
# MAIL_PASSWORD = 'passwordformail'
# MAIL_DEFAULT_SENDER = 'sender@gmail.com'

# The default user self registration role
AUTH_USER_REGISTRATION_ROLE = "Admin"

# When using OAuth Auth, uncomment to setup provider(s) info
# Google OAuth example:
OAUTH_PROVIDERS = [{
  'name':'google',
    'token_key':'access_token',
    'icon':'fa-google',
        'remote_app': {
            'api_base_url':'https://www.googleapis.com/oauth2/v2/',
            'client_kwargs':{
                'scope': 'email profile'
            },
            'access_token_url':'https://accounts.google.com/o/oauth2/token',
            'authorize_url':'https://accounts.google.com/o/oauth2/auth',
            'request_token_url': None,
            'client_id': '<MY_CLIENT_ID>',
            'client_secret': '<MY_CLIENT_SECRET>',
        }
}]
```

Will appreciate any help for this issue.

### What you expected to happen

Being able to connect to the web UI, and seeing the DAGs and everything.

### Operating System

Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

```
apache-airflow-providers-amazon==2.6.0
apache-airflow-providers-celery==2.1.0
apache-airflow-providers-cncf-kubernetes==3.0.1
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-google==6.3.0
apache-airflow-providers-http==2.0.2
apache-airflow-providers-imap==2.1.0
apache-airflow-providers-mysql==2.1.1
apache-airflow-providers-redis==2.0.1
apache-airflow-providers-sqlite==2.0.1
```

### Deployment

Other Docker-based deployment

### Deployment details

Using a custom Docker that is built from `python:3.7-buster`. Installing Airflow is done through pip install.
Our pip freeze is:
```
alembic==1.7.5
amqp==5.0.9
anyio==3.5.0
apache-airflow==2.2.3
apache-airflow-providers-amazon==2.6.0
apache-airflow-providers-celery==2.1.0
apache-airflow-providers-cncf-kubernetes==3.0.1
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-google==6.3.0
apache-airflow-providers-http==2.0.2
apache-airflow-providers-imap==2.1.0
apache-airflow-providers-mysql==2.1.1
apache-airflow-providers-redis==2.0.1
apache-airflow-providers-sqlite==2.0.1
apispec==3.3.2
argcomplete==1.12.3
asn1crypto==1.4.0
attrs==20.3.0
Authlib==0.15.5
Babel==2.9.1
bcrypt==3.2.0
beautifulsoup4==4.10.0
billiard==3.6.4.0
blinker==1.4
boto3==1.18.65
botocore==1.21.65
cached-property==1.5.2
cachetools==4.2.4
cattrs==1.5.0
celery==5.1.2
certifi==2021.10.8
cffi==1.15.0
charset-normalizer==2.0.10
click==7.1.2
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.2.0
clickclick==20.10.2
colorama==0.4.4
colorlog==4.8.0
commonmark==0.9.1
croniter==1.0.15
cryptography==36.0.1
decorator==5.1.1
defusedxml==0.7.1
dill==0.3.4
distlib==0.3.4
dnspython==2.2.0
docutils==0.16
email-validator==1.1.3
filelock==3.4.2
Flask==1.1.4
Flask-AppBuilder==3.4.3
Flask-Babel==2.0.0
Flask-Bcrypt==0.7.1
Flask-Caching==1.10.1
Flask-JWT-Extended==3.25.1
Flask-Login==0.4.1
Flask-OpenID==1.3.0
Flask-SQLAlchemy==2.5.1
flask-talisman==0.8.1
Flask-WTF==0.14.3
flower==1.0.0
google-ads==14.0.0
google-api-core==1.31.5
google-api-python-client==1.12.10
google-auth==1.35.0
google-auth-httplib2==0.1.0
google-auth-oauthlib==0.4.6
google-cloud-appengine-logging==1.1.0
google-cloud-audit-log==0.2.0
google-cloud-automl==2.6.0
google-cloud-bigquery==2.32.0
google-cloud-bigquery-datatransfer==3.5.0
google-cloud-bigquery-storage==2.11.0
google-cloud-bigtable==1.7.0
google-cloud-build==3.7.1
google-cloud-container==1.0.1
google-cloud-core==1.7.2
google-cloud-datacatalog==3.6.2
google-cloud-dataproc==3.2.0
google-cloud-dataproc-metastore==1.3.1
google-cloud-dlp==1.0.0
google-cloud-kms==2.10.1
google-cloud-language==1.3.0
google-cloud-logging==2.7.0
google-cloud-memcache==1.0.0
google-cloud-monitoring==2.8.0
google-cloud-os-login==2.5.1
google-cloud-pubsub==2.9.0
google-cloud-redis==2.5.1
google-cloud-secret-manager==1.0.0
google-cloud-spanner==1.19.1
google-cloud-speech==1.3.2
google-cloud-storage==1.44.0
google-cloud-tasks==2.7.2
google-cloud-texttospeech==1.0.1
google-cloud-translate==1.7.0
google-cloud-videointelligence==1.16.1
google-cloud-vision==1.0.0
google-cloud-workflows==1.5.0
google-crc32c==1.3.0
google-resumable-media==1.3.3
googleapis-common-protos==1.54.0
graphviz==0.19.1
grpc-google-iam-v1==0.12.3
grpcio==1.43.0
grpcio-gcp==0.2.2
gunicorn==20.1.0
h11==0.12.0
httpcore==0.13.7
httplib2==0.20.2
httpx==0.19.0
humanize==3.13.1
idna==3.3
importlib-metadata==4.10.1
importlib-resources==5.4.0
inflection==0.5.1
iso8601==1.0.2
isodate==0.6.1
itsdangerous==1.1.0
Jinja2==2.11.3
jmespath==0.10.0
json-merge-patch==0.2
jsonpath-ng==1.5.3
jsonschema==3.2.0
kombu==5.2.3
kubernetes==21.7.0
lazy-object-proxy==1.7.1
libcst==0.4.0
lockfile==0.12.2
lxml==4.7.1
Mako==1.1.6
Markdown==3.3.6
MarkupSafe==2.0.1
marshmallow==3.14.1
marshmallow-enum==1.5.1
marshmallow-oneofschema==3.0.1
marshmallow-sqlalchemy==0.26.1
mypy-extensions==0.4.3
mysql-connector-python==8.0.28
mysqlclient==2.1.0
ndg-httpsclient==0.5.1
nox==2020.12.31
numpy==1.21.5
oauthlib==3.1.1
openapi-schema-validator==0.1.6
openapi-spec-validator==0.3.2
packaging==21.3
pandas==1.3.5
pandas-gbq==0.14.1
pendulum==2.1.2
platformdirs==2.4.1
ply==3.11
prison==0.2.1
prometheus-client==0.12.0
prompt-toolkit==3.0.24
proto-plus==1.18.1
protobuf==3.19.3
psutil==5.9.0
py==1.11.0
pyarrow==6.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pydata-google-auth==1.3.0
Pygments==2.11.2
PyJWT==1.7.1
pyOpenSSL==21.0.0
pyparsing==3.0.7
pyrsistent==0.16.1
python-daemon==2.3.0
python-dateutil==2.8.2
python-nvd3==0.15.0
python-slugify==4.0.1
python3-openid==3.2.0
pytz==2021.3
pytzdata==2020.1
PyYAML==5.4.1
redis==3.5.3
redshift-connector==2.0.903
requests==2.27.1
requests-oauthlib==1.3.0
rfc3986==1.5.0
rich==11.0.0
rsa==4.8
s3transfer==0.5.0
scramp==1.4.1
setproctitle==1.2.2
six==1.16.0
sniffio==1.2.0
soupsieve==2.3.1
SQLAlchemy==1.3.24
SQLAlchemy-JSONField==1.0.0
sqlalchemy-redshift==0.8.9
SQLAlchemy-Utils==0.38.2
swagger-ui-bundle==0.0.9
tabulate==0.8.9
tenacity==8.0.1
termcolor==1.1.0
text-unidecode==1.3
tornado==6.1
typing-inspect==0.7.1
typing_extensions==4.0.1
unicodecsv==0.14.1
uritemplate==3.0.1
urllib3==1.26.8
vine==5.0.0
virtualenv==20.13.0
watchtower==2.0.1
wcwidth==0.2.5
websocket-client==1.2.3
Werkzeug==1.0.1
WTForms==2.3.3
zipp==3.7.0
```

This docker image is running on K8s (using an in-house system), and we are getting a domain pointing to the deployment pods.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
This is likely something wrong in in your k8s setup. This is not oauth  issue IMHO. You probably use "random" to generate [SECRET_KEY](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#secret-key) - and you have different secret keys in your UI and workers (make sure all your components have the same secret key configured). 

You have to make sure that the secret key is the same everywhere. I heartily recommend to use Airflow Helm Chart https://airflow.apache.org/docs/helm-chart/stable/index.html 
 and Production image: https://airflow.apache.org/docs/docker-stack/index.html instead of running your custom image / K8S deployment as many of similar problems with configurations are sorted out there, vetted by multiple people and tested.

I am turning that into discussion in case it does not help and in case you want to provide more information  (maybe my guess is incorrect).

**Reply by alexkruWix** [reply]
Thank you very much for responding for this. 
If possible, I have some followup questions:
* As mentioned, we set the `secret_key` config in the airflow.cfg file. We also tried setting the env vars. So  how can I check if I'm using a "random" value to generate the secret_key, if we explicitely set it in the aiflow.cfg?
* We currently run only 1 pod of webserver, and 1 pod of scheduler, and the value is the same in both (because we run the same docker image).

Unfortunately, as our system is built, we have limited access to the k8s deployment, so we can not run helm normally :( We will check why we don't use the official Docker images in the first place (probably legacy reason).

**Reply by potiuk** [reply]
Then my guess was wrong. Likely there is some problem with they way how the token gets somewhere stripped off by the redirection. I think  this it is not airflow but your ingress/proxy/firewall setting somewhere.  But I do not know unfortunately and it's likely you need to debug it in your system

The basic idea (and it's not really Airflow - it's Flask's CSRF protection) is that it generates a token (using the secret key) that then should be used when the redirection happens and the form that earlier was generated by Airflow is redirected back in the URL.

https://testdriven.io/blog/csrf-flask/

I am not 100% sure how it is done with teh oauth because it involves several redirections possibly, but this token - when generated once should be passed back when you redirect it back. Apparently the error is get is that this token is wrong - either it's been removed and should not be or (maybe - that's another option) the oauth specification does not require it, in whic case this might be a bug in Flask Application Builder (because Flask Application Builder is the one that provides Oauth authentication). Maybe the "url" you call should be exempt from CSRF protection?  Or maybe you need some extra parameters in your oauth configuration - not sure.

If you have one webserver only, then it is not likely it is "random secret".

What I can suggest is to debug it and see where the token is possibly tampered with, or if it does not work - you can open a discussion/create an issue in https://github.com/dpgaspar/flask-appbuilder/

**Reply by alexkruWix** [reply]
I will try to debug it, thank you! 

I have another question that I would like to clarify with you. In my way of debugging it, I tried to completely turn off the csrf protection (I know it's not recommended, but I just wanted to see stuff work), so I set the `WTF_CSRF_ENABLED` to `False` in my `webserver_config.py` file. To my surprise, I keep getting the same errors, although presumably csrf protection should be turned off. 
Does this makes any sense? Or does this issue is coming from the flask-builder itself, and it's not related to Airflow in the begin with? It's hard for me to follow, as I see that the errors is coming from the `views.py:666` file and I don't see the full stacktrace..

**Reply by potiuk** [reply]
That's what I'd do too :) 

It's strange - If you are sure thtat your change was effective (restarted webserver?), then it might also be slightly wrong in my diagnosis. This might actually be a CSRF protection in the Oauth itself, not the CSRF that airflow adds in the webserver - this actually is much more plausible explanation (as then indeed the SECRET_KEY of Airflow would not be used). 

Quickl search shows that you are not the only one with this error: https://github.com/lepture/authlib/issues/376 and this looks like an OAUTH (mis?) configuration issue? 

From past experience (and few people and projects where Oauth integration was needed) those kind of errors are pretty "normal" - oauth is built in with security in mind, so it gives as little clue as possible when things are wrong so that potential attacker cannot use it to increase the likelihood of breaking it. But maybe other parts of the system (Oauth server?) mihgt have some more information in their logs.

**Comment by KristiKovacs** [comment]
Just wondering if this was ever solved @alexkruWix ? I'm coming across the same issue.

**Reply by srp-pawar** [reply]
@alexkruWix , @KristiKovacs  
was  the error solved ?? 
I am also getting the same error 
while integrating WSO2 Identity server with superset OAUTH
please need some help to solve this error

Here is the error 
**superset_app          | 2024-01-16 07:25:59,280:ERROR:flask_appbuilder.security.views:Error authorizing OAuth access token: mismatching_state: CSRF Warning! State not equal in request and response.
superset_app          | 172.19.0.1 - - [16/Jan/2024:07:25:59 +0000] "GET /oauth-authorized/wso2?code=2daa86bc-78dc-34d0-b4f5-2b2b9b02d52d&state=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuZXh0IjpbIiJdfQ.9JJyGmzhaHKsTuNQ6BjjzG2E9z8pAzaDEeFNoV6Qiao&session_state=8ca1abbfc4f377c036e88a437c9a14e5ccad5045f81f0d101ea5b99d66bf72de.SZ8P2QTMsaqlpF96PWq0rg HTTP/1.1" 302 201 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"**
 
 and here is my superset_config file: 
 
import logging
import os

from celery.schedules import crontab
from flask_caching.backends.filesystemcache import FileSystemCache
from typing import Literal
logger = logging.getLogger()
SESSION_COOKIE_SAMESITE = "Strict"
ENABLE_PROXY_FIX = True
PUBLIC_ROLE_LIKE = "Gamma"
PUBLIC_ROLE_LIKE_GAMMA = True

FEATURE_FLAGS = {
    "EMBEDDED_SUPERSET": True
}
SQLALCHEMY_DATABASE_URI = 'postgresql://postgres:postgres@db:5432/superset'
ENABLE_CORS = True

TALISMAN_ENABLED = False
# ENABLE_CORS = True
HTTP_HEADERS={"X-Frame-Options":"ALLOWALL"} 
CSRF_ENABLED = False
CORS_OPTIONS = {
  'supports_credentials': True,
  'allow_headers': ['*'],
  'resources':['*'],
  'origins': ['*']
}
SECRET_KEY='76AqIdFq4U69MYuONDAYNtUUYesEQQ5LInmLsZzE66CRBOP1XYztxByq'

DATABASE_DIALECT = os.getenv("DATABASE_DIALECT")
DATABASE_USER = os.getenv("DATABASE_USER")
DATABASE_PASSWORD = os.getenv("DATABASE_PASSWORD")
DATABASE_HOST = os.getenv("DATABASE_HOST")
DATABASE_PORT = os.getenv("DATABASE_PORT")
DATABASE_DB = os.getenv("DATABASE_DB")

EXAMPLES_USER = os.getenv("EXAMPLES_USER")
EXAMPLES_PASSWORD = os.getenv("EXAMPLES_PASSWORD")
EXAMPLES_HOST = os.getenv("EXAMPLES_HOST")
EXAMPLES_PORT = os.getenv("EXAMPLES_PORT")
EXAMPLES_DB = os.getenv("EXAMPLES_DB")

SQLALCHEMY_DATABASE_URI = (
    f"{DATABASE_DIALECT}://"
    f"{DATABASE_USER}:{DATABASE_PASSWORD}@"
    f"{DATABASE_HOST}:{DATABASE_PORT}/{DATABASE_DB}"
)

SQLALCHEMY_EXAMPLES_URI = (
    f"{DATABASE_DIALECT}://"
    f"{EXAMPLES_USER}:{EXAMPLES_PASSWORD}@"
    f"{EXAMPLES_HOST}:{EXAMPLES_PORT}/{EXAMPLES_DB}"
)

REDIS_HOST = os.getenv("REDIS_HOST", "redis")
REDIS_PORT = os.getenv("REDIS_PORT", "6379")
REDIS_CELERY_DB = os.getenv("REDIS_CELERY_DB", "0")
REDIS_RESULTS_DB = os.getenv("REDIS_RESULTS_DB", "1")

RESULTS_BACKEND = FileSystemCache("/app/superset_home/sqllab")

CACHE_CONFIG = {
    "CACHE_TYPE": "RedisCache",
    "CACHE_DEFAULT_TIMEOUT": 300,
    "CACHE_KEY_PREFIX": "superset_",
    "CACHE_REDIS_HOST": REDIS_HOST,
    "CACHE_REDIS_PORT": REDIS_PORT,
    "CACHE_REDIS_DB": REDIS_RESULTS_DB,
}
DATA_CACHE_CONFIG = CACHE_CONFIG


class CeleryConfig:
    broker_url = f"redis://{REDIS_HOST}:{REDIS_PORT}/{REDIS_CELERY_DB}"
    imports = ("superset.sql_lab",)
    result_backend = f"redis://{REDIS_HOST}:{REDIS_PORT}/{REDIS_RESULTS_DB}"
    worker_prefetch_multiplier = 1
    task_acks_late = False
    beat_schedule = {
        "reports.scheduler": {
            "task": "reports.scheduler",
            "schedule": crontab(minute="*", hour="*"),
        },
        "reports.prune_log": {
            "task": "reports.prune_log",
            "schedule": crontab(minute=10, hour=0),
        },
    }


CELERY_CONFIG = CeleryConfig

FEATURE_FLAGS = {"ALERT_REPORTS": True}
ALERT_REPORTS_NOTIFICATION_DRY_RUN = True
WEBDRIVER_BASEURL = "http://superset:8088/"
WEBDRIVER_BASEURL_USER_FRIENDLY = WEBDRIVER_BASEURL

SQLLAB_CTAS_NO_LIMIT = True

#
# Optionally import superset_config_docker.py (which will have been included on
# the PYTHONPATH) in order to allow for local settings to be overridden
#
try:
    import superset_config_docker
    from superset_config_docker import *  # noqa

    logger.info(
        f"Loaded your Docker configuration at " f"[{superset_config_docker.__file__}]"
    )
except ImportError:
    logger.info("Using default Docker config...")
WTF_CSRF_ENABLED = False
SESSION_COOKIE_HTTPONLY = True 
SESSION_COOKIE_SECURE = False 


from flask_appbuilder.security.manager import AUTH_OAUTH
from custom_sso_security_manager import CustomSsoSecurityManager

CUSTOM_SECURITY_MANAGER = CustomSsoSecurityManager 

AUTH_TYPE = AUTH_OAUTH

AUTH_USER_REGISTRATION = True

AUTH_USER_REGISTRATION_ROLE = "Public"


OAUTH_PROVIDERS = [
    {
        'name':'wso2',
        'token_key':'access_token',
        'icon':'fa-address-card',
        'remote_app': {
            'client_id':'fpsr_uKduRtRtrgsD1x02ysJhzca',
            'client_secret':'ZX1qMVLxyP1BOmMc9sX5sR3NfQga',
            'api_base_url':'https://wso2:9443/oauth2',
            'client_kwargs':{
                'scope': 'read openid'
            },
            'jwks_uri':'https://wso2:9443/oauth2/jwks', # may be required to generate token
            'grant_type' : 'authorization_code',
            'access_token_url':'https://wso2:9443/oauth2/token',
            'authorize_url':'https://wso2:9443/oauth2/authorize',
            'access_token_headers':{    # Additional headers for calls to access_token_url
                'Authorization': 'Basic Base64EncodedClientIdAndSecret'
            },
        },
    }

]

SESSION_COOKIE_NAME='superset-session'

DEBUG = True
